agent/consul/gateways/controller_gateways.go

```diff
@@ -2,33 +2,202 @@ package gateways
 
 import (
 	"context"
-
-	"github.com/hashicorp/go-hclog"
-
 	"github.com/hashicorp/consul/agent/consul/controller"
-	"github.com/hashicorp/consul/agent/consul/fsm"
+	"github.com/hashicorp/consul/agent/consul/gateways/datastore"
 	"github.com/hashicorp/consul/agent/consul/state"
 	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/go-hclog"
+	"github.com/pkg/errors"
 )
 
 type apiGatewayReconciler struct {
-	fsm    *fsm.FSM
 	logger hclog.Logger
+	store  datastore.DataStore
 }
 
-func (r apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
-	return nil
-}
-
-func NewAPIGatewayController(fsm *fsm.FSM, publisher state.EventPublisher, logger hclog.Logger) controller.Controller {
+// NewAPIGatewayController returns a new APIGateway controller
+func NewAPIGatewayController(store datastore.DataStore, publisher state.EventPublisher, logger hclog.Logger) controller.Controller {
 	reconciler := apiGatewayReconciler{
-		fsm:    fsm,
 		logger: logger,
+		store:  store,
 	}
-	return controller.New(publisher, reconciler).Subscribe(
+	return controller.New(publisher, &reconciler).Subscribe(
 		&stream.SubscribeRequest{
 			Topic:   state.EventTopicAPIGateway,
 			Subject: stream.SubjectWildcard,
 		},
 	)
 }
+
+// Reconcile takes in a controller request and ensures this api gateways corresponding BoundAPIGateway exists and is
+// up to date
+func (r *apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+	metaGateway, err := r.initGatewayMeta(req)
+	if err != nil {
+		return err
+	} else if metaGateway == nil {
+		//delete meta gateway
+		r.logger.Info("cleaning up deleted gateway object", "request", req)
+		if err := r.store.Delete(&structs.BoundAPIGatewayConfigEntry{
+			Kind:           structs.BoundAPIGateway,
+			Name:           req.Name,
+			EnterpriseMeta: *req.Meta,
+		}); err != nil {
+			msg := "error cleaning up deleted gateway object"
+			r.logger.Error(msg, err)
+			return errors.Wrap(err, msg)
+		}
+		return nil
+	}
+
+	r.ensureBoundGateway(metaGateway)
+
+	r.logger.Debug("started reconciling gateway")
+
+	routes, err := r.retrieveAllRoutesFromStore()
+
+	if err != nil {
```

Comment:
#nit I like to keep these grouped together for readability
```suggestion
	routes, err := r.retrieveAllRoutesFromStore()
	if err != nil {
```